src/monitors/BranchMonitor.v3

```diff
@@ -7,7 +7,7 @@ class BmEntry(func: FuncDecl, pc: int, op: Opcode, numTargets: int) {
 	def taken = Array<u64>.new(numTargets);
 }
 
-class BranchMonitorState(module: Module) extends BytecodeVisitor {
+class BranchMonitorState(mm: ModuleMonitor) extends BytecodeVisitor {
```

Comment:
I think that makes sense. It would require changing the way `onFinish/onExit` is currently called though -- right now it's attached to the `Monitor` and all the registered monitors have their callbacks called at the appropriate times.

Would it make more sense to attach the `onStart`, `onMain` and `onFinish` callbacks to the `Module` instead, and change `onFinish` to take in a `Module` instead of an `Instance`? We can keep the `onExit` callbacks to `Monitor` since that seems to be a more global thing.

The alternative (that is, not changing the interface) is to have a dummy monitor that allows callbacks to be registered, and then registering that dummy monitor.

Reply:
The main thing I am trying to avoid is having to go look up the state for a module in the `onFinish/onExit` callbacks (or any callback). Maybe the `ModuleMonitor` can have methods to insert callbacks on these events, rather than methods to be overridden? Then the `BranchMonitorState` or others could just supply callbacks that implicitly close over themselves. Wdyt?

Reply:
I took a stab at doing this, not sure if it's what you had in mind. I think I can't make `BranchMonitorState` a direct subclass of `ModuleMonitor` because I still need its `BytecodeVisitor` functionality, and I had to change `monitors` from a immutable `List` to a `Vector` so `ModuleMonitor` is able to register itself after parsing.